### PR TITLE
Feature: allow custom width/height of screenshot via console

### LINKED ISSUE #8798: No "ERROR: " prefix in error message for "content select all"
## Version of OpenTTD
1.11.0-beta2

## Expected result
The output of "content select all" should be "ERROR: 'select all' is no longer supported since 1.11."

In other words, line 1837 here
https://github.com/OpenTTD/OpenTTD/blob/a4035af3375825ae277f1363c5d814ee55b92d7c/src/console_cmds.cpp#L1837
should instead be either of the following:
```c++
IConsoleError("'select all' is no longer supported since 1.11");
```
or (in case the `IConsoleError` method can't be used here)
```c++
IConsolePrintF(CC_ERROR, "ERROR: 'select all' is no longer supported since 1.11");
```

## Actual result
image
(Notice there's no "ERROR: " prefix next to the error message)

## Steps to reproduce
Run the following console commands while in the title screen:
```
content select all
kick
```

### PR DESCRIPTION
**Pull request by TrueBrain**
Used #8598 for inspiration and this PR is meant as alternative to that PR.

Fixes #8798
Yes, I sneaked that in this PR, I am sorry .. I could make a new PR out of it if you like.

## Motivation / Problem

For the titlegame competition I was looking for a good way to make screenshots of different resolutions in a single run in an automated way. #8598 allows this, but I noticed we could do better. It started out as renaming `res_x` to `width`, adding docstrings, and more of these things, but I soon figured out the console parsing code was just a bit too stiff for me. For example, you could only set the size of the screenshot for normal zoom, not for big zoom, without there really being a reason not to allow that.

So I rewrote how screenshot command is parsed, and made it into a token-system. You can now mix and match different modes. Also, rewrote most of the implementation to be a bit more robust and future-proof, in my opinion of course.

## Description

```
Reworked how the screenshot command works while keeping it backwards
compatible. It can now more freely understand arguments, and has
the ability to make SC_DEFAULTZOOM screenshots.
```

**BONUS** you can now also make heightmap screenshots!

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/screenshot.cpp [hunk 1]
```diff
@@ -710,13 +710,17 @@ static bool MakeSmallScreenshot(bool crashlog)
...
+ * @param height the height of the screenshot, or 0 for current viewport height (needs to be 0 with SC_VIEWPORT, SC_CRASHLOG, and SC_WORLD).
  * @param[out] vp Result viewport
  */
-void SetupScreenshotViewport(ScreenshotType t, Viewport *vp)
+void SetupScreenshotViewport(ScreenshotType t, Viewport *vp, uint32 width, uint32 height)
 {
 	switch(t) {
 		case SC_VIEWPORT:
 		case SC_CRASHLOG: {
+			assert(width == 0 && height == 0);
```
Comment (glx22) [inline src/screenshot.cpp:722]: Assert on user input is not the best thing to do. Especially as asserts can be disabled.

Comment (TrueBrain) [reply src/screenshot.cpp:722]: But it is not based on user input. It is enforcing a rule described in the doxygen comments.

The only ones calling this function are `MakeLargeWorldScreenshot` and `MakeScreenshotWithConfirm`. The latter always calls it with 0 (and the asserts confirm they do), and the first only allows non-zero with `SC_ZOOMEDIN` and `SC_DEFAULTZOOM`.

This is also the reason I added the asserts, to make sure not someone in a year goes: owh, I want to make the `SC_VIEWPORT` size changeable. He will immediately hit the assert, reminding him that the code is not made to do that.

At least, that is my reasoning :) Open for suggestions and alternatives, of course. But this is not user input :)

Comment (TrueBrain) [reply src/screenshot.cpp:722]: What did trigger me with this comment, that I can just tell the user when he tries `size` in anything not `normal` or `big` it won't do anything. Might be nicer :D

Comment (glx22) [reply src/screenshot.cpp:722]: Ah yes, now (on real browser) I see. Reviewing in the mobile app is not the best idea ;)

File: src/console_cmds.cpp [hunk 2]
```diff
@@ -1868,7 +1868,7 @@ DEF_CONSOLE_CMD(ConContent)
 			 * to download every available package on BaNaNaS. This is not in
 			 * the spirit of this service. Additionally, these few people were
 			 * good for 70% of the consumed bandwidth of BaNaNaS. */
-			IConsolePrintF(CC_ERROR, "'select all' is no longer supported since 1.11");
+			IConsoleError("'select all' is no longer supported since 1.11");
```
Comment (James103) [inline src/console_cmds.cpp:1871]: This commit should probably go in its own separate PR (that mentions that it fixes #8798).

**Comment by nielsmh** [comment]
What does the "viewport" mode of the screenshot command do on a dedicated server?

**Comment by TrueBrain** [comment]
> 
> 
> What does the "viewport" mode of the screenshot command do on a dedicated server?

Surprisingly uneventful.

If you try it without manually selecting a blitter, you just get an error.
But if you do! Well .. still nothing exciting happens. It just makes a screenshot like it is a normal game:

screenshot

So yeah ... not really exciting, sorry :P

What is more exciting, is what happens if you make a lot of screenshots .. you get a gif like this:

test

Dedicated servers don't run certain graphical routines :P

**Comment by glx22** [comment]
Haha, that's why you apply some patches for the title game contest

**Review by glx22 — APPROVED** [review]
`screenshot` command looks cleaner that way.
